### ISSUE TITLE
removing train parameter on collate_fn for huggingface's Trainer

### ISSUE THREAD

**Issue by hadariru** [ariG23498/ft-pali-gemma#1]
Thank you for sharing the finetuning code on paligemma, this really helps me!

I have a question for collate_fn.

Seeing the example on huggingface and yours, I would like to know if it is possible to use Trainer from huggingface instead of using the normal loop like yours.

I am not sure on how to remove the train=True parameter on your collate_fn function

```
    collate_fn_trainer = partial(collate_fn, 
                                 image_title="image", 
                                 prompt="Detect license plate.", 
                                 suffix_title="label_for_paligemma", 
                                 processor=processor, 
                                 device=device, 
                                 train=True) -> how to 
    trainer = Trainer(
            model=model,
            train_dataset=train_dataset,
            eval_dataset=validation_dataset,
            data_collator=collate_fn_trainer,
            args=args
            )
```

**Comment by ariG23498** [comment]
Hey!

Thanks for your interest in this work :)

I am not very familiar with the `Trainer` module of hugging face. Could you give me a colab notebook to reproduce the training loop as you might like it to be? It would be easier for me to look into it and comment.

**Comment by hadariru** [comment]
this is the colab version
I created the dataset using your script. Mine was not a notebook code. So I only copied relevant parts of my code into a notebook code


https://colab.research.google.com/drive/1AOfWNH1CWgiJuz8-xTMfcLuFFlgFIBLd?usp=sharing
at the 4th cell, you can see that dataset is passed into trainer, right now i think all of the dataset would have train=True since they are using the same collator_fn

**Comment by ariG23498** [comment]
Thank you for sharing the colab notebook.

As I understand, you would like to know whether the `train` argument in the `collate_fn` should be `True` or `False`. If so, then going into the docs of `Trainer` I see that
```
data_collator (DataCollator, optional) — The function to use to form a batch from a list of elements of 
train_dataset or eval_dataset. Will default to [default_data_collator()]
(https://huggingface.co/docs/transformers/v4.41.3/en/main_classes/data_collator#transformers.default_data_collator) if 
no tokenizer is provided, an instance of [DataCollatorWithPadding]
(https://huggingface.co/docs/transformers/v4.41.3/en/main_classes/data_collator#transformers.DataCollatorWithPadding
) otherwise.
```
this means that the `data_collator` which is `collate_fn` here is used for training and evaluation datasets. With this knowledge I think the `train` argument should be `True`. When we test our model, we need to switch the `train` argument to `False` so that the model generates the bounding boxes.

Hope this helps. Also, if I was unable to understand the question, please feel free to make the question more verbose for me so that I can help you further. 😄

**Comment by hadariru** [comment]
Sorry my question wasn't very clear.

when training, `collate_fn` must be `True`, while for testing, the `collate_fn` must be `False`
On your script, I can see that you are using dataloader, so that different collate_fn function can be used for each parts.

But in hf's Trainer case, I only have 1 `collate_fn` which is created before the Trainer class

```
collate_fn_trainer = partial(collate_fn,
                                image_title="image",
                                prompt="Detect table.",
                                suffix_title="paligemma_label",
                                processor=processor,
                                device=device,
                                train=model.training) # create collate_fn for trainer, but this is True
trainer = Trainer(
        model=model,
        train_dataset=train_dataset,
        eval_dataset=validation_dataset,
        data_collator=collate_fn_trainer, # collate_fn_trainer train is always True, how to change into false?
        args=args
        )
```

I am not sure on how to change the train value on `collate_fn` into `False` for each dataset (train_dataset and eval_dataset)

**Comment by BramVanroy** [comment]
I would probably create your own subclass of the Trainer, and override its data loader methods:

https://github.com/huggingface/transformers/blob/9454f437b0ba87e29cae9706cee6be6bb88d28d6/src/transformers/trainer.py#L865

https://github.com/huggingface/transformers/blob/9454f437b0ba87e29cae9706cee6be6bb88d28d6/src/transformers/trainer.py#L922

https://github.com/huggingface/transformers/blob/9454f437b0ba87e29cae9706cee6be6bb88d28d6/src/transformers/trainer.py#L969

**Comment by hadariru** [comment]
Not sure if this is the correct way, but as @BramVanroy said, I subclass it and changed only the data_collator function

```

class PaliGemmaImageTrainer(Trainer):
    def get_train_dataloader(self) -> DataLoader:
        """
        Returns the training [`~torch.utils.data.DataLoader`].

        Will use no sampler if `train_dataset` does not implement `__len__`, a random sampler (adapted to distributed
        training if necessary) otherwise.

        Subclass and override this method if you want to inject some custom behavior.
        """
        if self.train_dataset is None:
            raise ValueError("Trainer: training requires a train_dataset.")

        train_dataset = self.train_dataset
        data_collator = partial(self.data_collator, train=True)
        if is_datasets_available() and isinstance(train_dataset, datasets.Dataset):
            train_dataset = self._remove_unused_columns(train_dataset, description="training")
        else:
            data_collator = self._get_collator_with_removed_columns(data_collator, description="training")

        dataloader_params = {
            "batch_size": self._train_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(train_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_train_sampler()
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["worker_init_fn"] = seed_worker
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        return self.accelerator.prepare(DataLoader(train_dataset, **dataloader_params))
    
    def get_eval_dataloader(self, eval_dataset: Optional[Dataset] = None) -> DataLoader:
        """
        Returns the evaluation [`~torch.utils.data.DataLoader`].

        Subclass and override this method if you want to inject some custom behavior.

        Args:
            eval_dataset (`torch.utils.data.Dataset`, *optional*):
                If provided, will override `self.eval_dataset`. If it is a [`~datasets.Dataset`], columns not accepted
                by the `model.forward()` method are automatically removed. It must implement `__len__`.
        """
        if eval_dataset is None and self.eval_dataset is None:
            raise ValueError("Trainer: evaluation requires an eval_dataset.")

        # If we have persistent workers, don't do a fork bomb especially as eval datasets
        # don't change during training
        if hasattr(self, "_eval_dataloader") and self.args.dataloader_persistent_workers:
            return self.accelerator.prepare(self._eval_dataloader)
        eval_dataset = eval_dataset if eval_dataset is not None else self.eval_dataset
        data_collator = partial(self.data_collator, train=False)


        if is_datasets_available() and isinstance(eval_dataset, datasets.Dataset):
            eval_dataset = self._remove_unused_columns(eval_dataset, description="evaluation")
        else:
            data_collator = self._get_collator_with_removed_columns(data_collator, description="evaluation")

        dataloader_params = {
            "batch_size": self.args.eval_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(eval_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_eval_sampler(eval_dataset)
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        # accelerator.free_memory() will destroy the references, so
        # we need to store the non-prepared version
        eval_dataloader = DataLoader(eval_dataset, **dataloader_params)
        if self.args.dataloader_persistent_workers:
            self._eval_dataloader = eval_dataloader

        return self.accelerator.prepare(eval_dataloader)
    
    def get_test_dataloader(self, test_dataset: Dataset) -> DataLoader:
        """
        Returns the test [`~torch.utils.data.DataLoader`].

        Subclass and override this method if you want to inject some custom behavior.

        Args:
            test_dataset (`torch.utils.data.Dataset`, *optional*):
                The test dataset to use. If it is a [`~datasets.Dataset`], columns not accepted by the
                `model.forward()` method are automatically removed. It must implement `__len__`.
        """
        data_collator = partial(self.data_collator, train=False)

        if is_datasets_available() and isinstance(test_dataset, datasets.Dataset):
            test_dataset = self._remove_unused_columns(test_dataset, description="test")
        else:
            data_collator = self._get_collator_with_removed_columns(data_collator, description="test")

        dataloader_params = {
            "batch_size": self.args.eval_batch_size,
            "collate_fn": data_collator,
            "num_workers": self.args.dataloader_num_workers,
            "pin_memory": self.args.dataloader_pin_memory,
            "persistent_workers": self.args.dataloader_persistent_workers,
        }

        if not isinstance(test_dataset, torch.utils.data.IterableDataset):
            dataloader_params["sampler"] = self._get_eval_sampler(test_dataset)
            dataloader_params["drop_last"] = self.args.dataloader_drop_last
            dataloader_params["prefetch_factor"] = self.args.dataloader_prefetch_factor

        # We use the same batch_size as for eval.
        return self.accelerator.prepare(DataLoader(test_dataset, **dataloader_params))
```

**Comment by hadariru** [comment]
Is there anything looking wrong?
while training I can see that loss is decreasing, from 6 to around 2
But I am not seeing any detection results from the model.

```

    print("Before fine tuning")
    # run model generation before fine tuning
    test_batch = next(iter(test_dataloader))
    with torch.inference_mode():
        generation = model.generate(**test_batch, max_new_tokens=100, do_sample=False)
        decoded = processor.batch_decode(generation, skip_special_tokens=True)
        
    for index in range(object_detection_config.BATCH_SIZE):
        image = test_batch["pixel_values"][index].permute(1, 2, 0).cpu().float()
        element = decoded[index]
        detection_string = element.split("\n")[1]
        objects = extract_objects(detection_string, 224, 224, unique_labels=False)
        print(objects)
        draw_bbox(image, objects)
        
    from transformers import TrainingArguments
    args=TrainingArguments(
                output_dir="./output",
                num_train_epochs=20,
                remove_unused_columns=False,
                per_device_train_batch_size=4,
                gradient_accumulation_steps=4,
                warmup_steps=2,
                learning_rate=2e-5,
                weight_decay=1e-6,
                adam_beta2=0.999,
                logging_steps=100,
                optim="adamw_torch",
                save_strategy="steps",
                save_steps=1000,
                push_to_hub=False,
                save_total_limit=1,
                bf16=True,
                report_to=["tensorboard"],
                dataloader_pin_memory=False,
            
            )
    collate_fn_trainer = partial(collate_fn, 
                                 image_title="image", 
                                 prompt="Detect table.", 
                                 suffix_title="paligemma_label", 
                                 processor=processor, 
                                 device=device)
    trainer = PaliGemmaImageTrainer(
            model=model,
            train_dataset=train_dataset,
            eval_dataset=validation_dataset,
            data_collator=collate_fn_trainer,
            args=args
            )
    
    print("Build trainer with on device:", args.device, "with n gpus:", args.n_gpu)
    trainer.train()
    
    
    print("After fine tuning")
    test_batch = next(iter(test_dataloader))
    with torch.inference_mode():
        generation = trainer.model.generate(**test_batch, max_new_tokens=100, do_sample=False)
        decoded = processor.batch_decode(generation, skip_special_tokens=True)
        
    for index in range(object_detection_config.BATCH_SIZE):
        image = test_batch["pixel_values"][index].permute(1, 2, 0).cpu().float()
        element = decoded[index]
        detection_string = element.split("\n")[1]
        objects = extract_objects(detection_string, 224, 224, unique_labels=False)
        print(objects)
        draw_bbox(image, objects)
    
    
```
